### PR TITLE
Refactor FXIOS-8408 [v125] Adapt monorepo migration scripts to iOS

### PR DESCRIPTION
**Pull request by isabelrios**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-8408)

This PR adds scripts that were used to migrate android projects to a monorepo.

We have changed the references to android projects with firefox and focus ios when needed in different scripts and Json files, and also there are some changes related to the release branches names.

Aditionally, we have removed the regexes as they were specifically for android matching between projects.

We have created this document with a description of the scripts, the steps to run it, as well as some issues we had to fix when running this locally in case someone finds the same. 

https://docs.google.com/document/d/1WUOjC-TOPV4Xwj0BjKQVScCyzaiHEtLhyXYpkyJ3wbc/edit#heading=h.uvgimr49ls8z

This is an example of the new focus-ios-prep branch automatically created after running the migration script, sync-focus.sh
 https://github.com/isabelrios/firefox-ios/commits/focus-ios-prep/

## :pencil: Checklist
You have to check all boxes before merging
- [ ] Filled in the above information (tickets numbers and description of your work)
- [ ] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed I updated documentation / comments for complex code and public methods

### CODE REVIEW

**Comment by isabelrios** [comment]
@JohanLorenzo, please let me ping you to try to answer a few questions we have once we see our focus-ios-prep branch created:
- The branch has a rewritten history including focus repo history, does that look good to you?
- If that's correct, what would be the expected  status once this is merged?

There might be more questions, but wanted to leave those for now, thank you for your help! 🙏

**Comment by JohanLorenzo** [comment]
> I'm wondering if we could use https://github.com/mozilla-releng/staging-firefox-ios or something similar to try the big merge in a different place first rather than our current main branch.. how was that done on android?

We could use it but I'm not sure what you would like to accomplish here. Without using a staging repo, you can visualize what the merge is going to look like at: https://github.com/mozilla-mobile/firefox-ios/compare/main...isabelrios:firefox-ios:focus-ios-prep

Then, I'd recommend creating a PR from there. This way, you'll be able to do the big merge through the regular PR process. 

On Android, we used the staging repository because of Taskcluster. We had dozens of checks to perform on a repo that runs all the CI/CD on Taskcluster. I'm not sure if we have similar needs here but I'm happy to discuss this 🙂 


>     * The branch has a rewritten history including focus repo history, does that look good to you?

Looking at the link above:
 * the number of commits looks good
 * authorship was preserved 
 * issues and pull requests were correctly identified and got the right links (e.g. https://github.com/isabelrios/firefox-ios/commit/feeaa4314a6997667eb0d1ca50adcf5c2d7c4826) 
 * files are in the `focus-ios` subdirectory - please make sure this is the name for you all.
 

It seems the prefix `[focus-ios]` wasn't added to each commit message. That's the only issue I see so far.

>     * If that's correct, what would be the expected  status once this is merged?

I'm sorry, I'm not sure I understand. Could you expand on this question?

**Comment by JohanLorenzo** [comment]
One more thing: could create 2 commits for this current PR? The first commit contains the original files in https://github.com/mozilla-mobile/firefox-android/tree/aa1a25c43eba010b7a82a6550a99e41af043adf6/monorepo-migration. The second contains the changes to make it work on focus-ios. That will help reviewing this PR 🙂

**Review by JohanLorenzo — COMMENTED** [review]
One more thing: could create 2 commits for this current PR? The first commit contains the original files in https://github.com/mozilla-mobile/firefox-android/tree/aa1a25c43eba010b7a82a6550a99e41af043adf6/monorepo-migration. The second contains the changes to make it work on focus-ios. That will help reviewing this PR 🙂

**Comment by isabelrios** [comment]
> It seems the prefix [focus-ios] wasn't added to each commit message. That's the only issue I see so far.

Oh right,  we will look into this.

> > If that's correct, what would be the expected status once this is merged?

> I'm sorry, I'm not sure I understand. Could you expand on this question?

Just to double check that we would see in main that new commits history, as it would be in the PR... 

I forgot to ask about the release branches... I guess we would need one for each one (focus-ios-prep-123 and focus-ios-prep-124)? Note that they have different names as the firefox release branch..

Thank your for your replies, very helpful @JohanLorenzo !

**Comment by JohanLorenzo** [comment]
> Just to double check that we would see in main that new commits history, as it would be in the PR...

Got it! In this case, you can create a PR from and to your own fork. This way, you can merge the PR and see the final result on your own fork. That's the fastest way to see results 🙂 

> I forgot to ask about the release branches... I guess we would need one for each one (focus-ios-prep-123 and focus-ios-prep-124)? Note that they have different names as the firefox release branch..

Yup! For each release branch, you will create a PR. You can look at what they're going to look like by doing the same as the `main` branch.

**Comment by AaronMT** [comment]
> We could use it but I'm not sure what you would like to accomplish here. Without using a staging repo, you can visualize what the merge is going to look like at: [main...isabelrios:firefox-ios:focus-ios-prep](https://github.com/mozilla-mobile/firefox-ios/compare/main...isabelrios:firefox-ios:focus-ios-prep)

Looking at the above big diff, just double checking if ascending commit order (e.g, first commit 2015) shown on Github correct?

**Comment by lmarceau** [comment]
Thank you all for the explanations about git history. I have one point to ask about the structure of the folder. I pulled in the [`focus-ios-prep`](https://github.com/isabelrios/firefox-ios/commits/focus-ios-prep/) branch and there's currently the following structure under the `focus-ios` folder.

Screenshot 2024-02-15 at 2 06 29 PM

1. How are we going to handle the duplicated files/files that are shared between the two projects? We should handle this before the merge of this branch right?
2. There's a second `focus-ios` subdirectory inside the first `focus-ios` directory. I think there should only be one of those (so right under the first `focus-ios` directory we should have the code related folders). Again, we'll need to handle this change before the big merge right?
Thank you!

**Comment by isabelrios** [comment]
> Thank you all for the explanations about git history. I have one point to ask about the structure of the folder. I pulled in the [`focus-ios-prep`](https://github.com/isabelrios/firefox-ios/commits/focus-ios-prep/) branch and there's currently the following structure under the `focus-ios` folder.
> 
> 
>     1. How are we going to handle the duplicated files/files that are shared between the two projects? We should handle this before the merge of this branch right?

I open this jira ticket to work on that:
https://mozilla-hub.atlassian.net/browse/FXIOS-8446

For those files that are in both repo we could use the ones in Firefox for now as they are. If I'm not mistaken, the only two actually shared that could break the app / ci in the root directore are:

- bitrise.yml file: we can ignore it in focus since the content of it has been already added to the one in firefox-ios repo. We can only have one for the two apps.
- swiftlint file: we are going to have many issues in focus since we have not been applying same rules there, so my suggestion is to add focus-ios/ path to be ignored and enable rules step by step there or fix the errors in focus repo while we work on these merge scripts. Whatever sounds better for you..

>     2. There's a second `focus-ios` subdirectory inside the first `focus-ios` directory. I think there should only be one of those (so right under the first `focus-ios` directory we should have the code related folders). Again, we'll need to handle this change before the big merge right?

You are right, all focus repo is moved into a focus-ios folder to do the  migration, we don't need that extra level. All files in that folder should be moved to the firefox-ios repo as they are, not insed that folder. I guess the easiest way is to create it as it is and then extract all the files after the merge but I may be mistaken missing something that could be better...

Let's talk about these topics.

**Comment by isabelrios** [comment]
> One more thing: could create 2 commits for this current PR? The first commit contains the original files in https://github.com/mozilla-mobile/firefox-android/tree/aa1a25c43eba010b7a82a6550a99e41af043adf6/monorepo-migration. The second contains the changes to make it work on focus-ios. That will help reviewing this PR 🙂

Actually, I have created a PR with the original files, then I can rebase and see the diff... 
https://github.com/mozilla-mobile/firefox-ios/pull/18756
Hope that works

File: monorepo-migration/generate-replace-message-expressions.py
Comment (isabelrios) [inline monorepo-migration/generate-replace-message-expressions.py:17]: Looks like this should be firefox-ios instead... https://github.com/mozilla-mobile/firefox-android/blob/94f8a7e8cc5a18fc7a1c9d10e647fd15a79ac952/monorepo-migration/generate-replace-message-expressions.py#L17C1-L17C20

File: monorepo-migration/sync-focus.sh
Comment (JohanLorenzo) [inline monorepo-migration/sync-focus.sh:1]: Nit: let's rename the file in a separate commit (which could be part of this PR still). It'll make the diff easier to review.

Comment (isabelrios) [reply monorepo-migration/sync-focus.sh:1]: Just realized and did that :) I will remove this sync-focus.sh and rename it later on

File: monorepo-migration/sync-fenix.sh
Comment (JohanLorenzo) [inline monorepo-migration/sync-fenix.sh:62]: Nit: No need to comment out this line anymore since we're already skipping this function in: 

https://github.com/mozilla-mobile/firefox-ios/pull/18727/files#diff-26f75150de3654539f07404abeb474003f616baa09620e854e587bfc611d1a85R128

File: monorepo-migration/generate-replace-message-expressions.py
Comment (JohanLorenzo) [reply monorepo-migration/generate-replace-message-expressions.py:17]: I think changing this line will fix: 

> It seems the prefix `[focus-ios]` wasn't added to each commit message. That's the only issue I see so far.

Let me know if it doesn't

Comment (isabelrios) [reply monorepo-migration/generate-replace-message-expressions.py:17]: Actually, when I change this to firefox-ios, the commit history for focus does not work, the commits are linked to old firefox-ios PRs... not sure why yet...

Comment (JohanLorenzo) [reply monorepo-migration/generate-replace-message-expressions.py:17]: Oh you're right! On Android, it was changed to `firefox-android` for the sake of the migration from the `firefox-android` repo to `mozilla-central` https://github.com/mozilla-mobile/firefox-android/commit/fb86715761d0f3de99c78076612c860687e9faee

Thus, we should stick to `focus-ios`.

Regarding the prefix of the commit message, these lines should do the trick: 

https://github.com/mozilla-mobile/firefox-android/commit/fb86715761d0f3de99c78076612c860687e9faee#diff-99a51aa3f36b7e3ffdd96456612906d550331ee10c456d7d7868135cef17e46fL38-L40

**Review by JohanLorenzo — APPROVED** [review]
Looks all good to me! I cloned your fork and ran some more checks:
 * the URLs to pulls/issues are the correct ones
 * `[focus-ios]` now prefixes the commit messages
 * the number of total commits looks sane

I didn't check the release branch but we can figure that out later. 

I leave it to other to check if the name of the new folder is the expected one. 

Great job in excavating and restoring this script! 👍

File: monorepo-migration/generate-replace-message-expressions.py
Comment (lmarceau) [inline monorepo-migration/generate-replace-message-expressions.py:60]: Is this a TBD? I had understood from a previous comment we might not need those at all?

Comment (isabelrios) [reply monorepo-migration/generate-replace-message-expressions.py:60]: Actually I have added a regex that is working now, so I can remove it or change it as "add more regexes if needed"...

**Review by lmarceau — APPROVED** [review]

**Pull request merged**
